@jordanlewis @rohany this can be given another look now.

Updated the tests to "wrap" the error, here's a snippet

```    
--- FAIL: TestDDLAnalysis/ddl_analysis (0.49s)
        datadriven.go:165: 
            testdata/ddl_analysis/ddl_analysis:41: CREATE ROLE roleb
            expected:
            7
            
            found:
            6
        ddl_round_trip_test.go:92: statement: CREATE ROLE roleb
        ddl_round_trip_test.go:95: NOTE: If the count (number of kv batch operations) increase is expected or insignificant or the count decreases, update the test counts.
        ddl_round_trip_test.go:98: The rewrite option can be used to overwrite all the counts.
        ddl_round_trip_test.go:99:      0.000ms      0.000ms    === operation:CREATE ROLE roleb client:127.0.0.1:62824 hostssl: node:1 sb:1 user:root
                 0.052ms      0.052ms    event:sql/conn_executor_exec.go:500 [n1,client=127.0.0.1:62824,hostssl,user=root] planning ends
                 0.072ms      0.019ms    event:sql/conn_executor_exec.go:500 [n1,client=127.0.0.1:62824,hostssl,user=root] checking distributability
                 0.080ms      0.008ms    event:sql/exec_util.go:917 [n1,client=127.0.0.1:62824,hostssl,user=root] query not supported for distSQL: unsupported node```

--- a/pkg/kv/kvclient/kvcoord/txn_coord_sender.go
+++ b/pkg/kv/kvclient/kvcoord/txn_coord_sender.go
@@ -32,7 +32,8 @@ import (
 )
 
 const (
-	opTxnCoordSender = "txn coordinator send"
+	// OpTxnCoordSender represents a txn coordinator send operation.
+	OpTxnCoordSender = "txn coordinator send"
 )
 
 // txnState represents states relating to whether an EndTxn request needs
@@ -468,7 +469,7 @@ func (tc *TxnCoordSender) Send(
 
 	startNs := tc.clock.PhysicalNow()
 
-	ctx, sp := tc.AnnotateCtxWithSpan(ctx, opTxnCoordSender)
+	ctx, sp := tc.AnnotateCtxWithSpan(ctx, OpTxnCoordSender)
 	defer sp.Finish()
 
 	// Associate the txnID with the trace.

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -252,6 +252,18 @@ func (ex *connExecutor) execStmtInOpenState(
 		}()
 	}
 
+	if ex.server.cfg.TestingKnobs.WithStatementTrace != nil {
+		tr := ex.server.cfg.AmbientCtx.Tracer
+		var sp opentracing.Span
+		ctx, sp = tracing.StartSnowballTrace(ctx, tr, stmt.SQL)
+		// TODO(radu): consider removing this if/when #46164 is addressed.
+		p.extendedEvalCtx.Context = ctx
+
+		defer func() {
+			ex.server.cfg.TestingKnobs.WithStatementTrace(sp, stmt.SQL)
+		}()
+	}
+
 	if ex.sessionData.StmtTimeout > 0 {
 		timeoutTicker = time.AfterFunc(
 			ex.sessionData.StmtTimeout-timeutil.Since(ex.phaseTimes[sessionQueryReceived]),

--- a/pkg/sql/ddl_round_trip_test.go
+++ b/pkg/sql/ddl_round_trip_test.go
@@ -1,0 +1,176 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql_test
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/testutils/diagutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/datadriven"
+	"github.com/cockroachdb/errors"
+	"github.com/opentracing/opentracing-go"
+)
+
+// TestDDLAnalysis runs the datadriven DDL analysis tests. These tests live in
+// the sql/testdata/ddl_analysis folder. The test allows us to approximate the
+// number of round trips a DDL statement takes by counting kv operations.
+// This allows us to catch DDL statements which are slow on geo-distributed
+// clusters simply due to unnecessary round trips.
+//
+// Supported directives:
+//
+//  - exec
+//
+//    Executes SQL statements against the database. Outputs no results on
+//    success. In case of error, outputs the error message.
+//
+//  - count
+//    Executes a SQL statement and counts the number of kv batch request
+//    operations that are made for the statement.
+
+type countWithTrace struct {
+	count int
+	trace string
+}
+
+func TestDDLAnalysis(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+
+	datadriven.Walk(t, "testdata/ddl_analysis", func(t *testing.T, path string) {
+		diagSrv := diagutils.NewServer()
+		defer diagSrv.Close()
+		var stmtToKvBatchRequests sync.Map
+
+		beforePlan := func(sp opentracing.Span, stmt string) {
+			if _, ok := stmtToKvBatchRequests.Load(stmt); ok {
+				sp.Finish()
+				trace := tracing.GetRecording(sp)
+				tc := countWithTrace{
+					count: countKvBatchRequestsInRecording(trace),
+					trace: trace.String(),
+				}
+				stmtToKvBatchRequests.Store(stmt, tc)
+			}
+		}
+
+		params := base.TestServerArgs{
+			Knobs: base.TestingKnobs{
+				SQLExecutor: &sql.ExecutorTestingKnobs{
+					WithStatementTrace: beforePlan,
+				},
+			},
+		}
+
+		s, sqlConn, _ := serverutils.StartServer(t, params)
+		defer s.Stopper().Stop(ctx)
+
+		// This allows us to log the trace of the failed test case.
+		// The trace we want to output will be for the last test case in the case of
+		// failure since a failing test case will stop the test.
+		var lastRunCountStmt string
+		defer func() {
+			if t.Failed() {
+				t.Logf("statement: %s", lastRunCountStmt)
+				load, _ := stmtToKvBatchRequests.Load(lastRunCountStmt)
+				tc := load.(countWithTrace)
+				t.Log("NOTE: If the change is not a significant increase, " +
+					"you can update the expected count without further investigation.")
+				t.Log("The rewrite option can be used to overwrite all the counts.")
+				t.Log(tc.trace)
+			}
+		}()
+
+		datadriven.RunTest(t, path, func(t *testing.T, td *datadriven.TestData) string {
+			switch td.Cmd {
+			case "exec":
+				_, err := sqlConn.Exec(td.Input)
+				if err != nil {
+					if errors.HasAssertionFailure(err) {
+						td.Fatalf(t, "%+v", err)
+					}
+					return fmt.Sprintf("error: %v\n", err)
+				}
+				return ""
+
+			case "count":
+				stmtToKvBatchRequests.Store(td.Input, countWithTrace{count: -1})
+				lastRunCountStmt = td.Input
+				_, err := sqlConn.Exec(td.Input)
+				if err != nil {
+					if errors.HasAssertionFailure(err) {
+						td.Fatalf(t, "%+v", err)
+					}
+					return fmt.Sprintf("error: %v\n", err)
+				}
+
+				out, _ := stmtToKvBatchRequests.Load(td.Input)
+				tc := out.(countWithTrace)
+				count := tc.count
+				if count == -1 {
+					t.Fatalf(
+						"could not find number of round trips for statement: %s",
+						td.Input,
+					)
+				}
+
+				return fmt.Sprintf("%d", count)
+
+			default:
+				td.Fatalf(t, "unknown command %s", td.Cmd)
+				return ""
+			}
+		})
+	})
+}
+
+// count the number of KvBatchRequests inside a recording, this is done by
+// counting each "txn coordinator send" operation.
+func countKvBatchRequestsInRecording(r tracing.Recording) int {
+	root := r[0]
+
+	// Find the topmost "flow" span to start traversing from.
+	for _, sp := range r {
+		if sp.ParentSpanID == root.SpanID && sp.Operation == "flow" {
+			return countKvBatchRequestsInSpan(r, sp)
+		}
+	}
+
+	return countKvBatchRequestsInSpan(r, root)
+}
+
+func countKvBatchRequestsInSpan(r tracing.Recording, sp tracing.RecordedSpan) int {
+	count := 0
+	// Count the number of OpTxnCoordSender operations while traversing the
+	// tree of spans.
+	if sp.Operation == kvcoord.OpTxnCoordSender {
+		count++
+	}
+
+	for _, osp := range r {
+		if osp.ParentSpanID != sp.SpanID {
+			continue
+		}
+		count += countKvBatchRequestsInSpan(r, osp)
+	}
+
+	return count
+}

--- a/pkg/sql/exec_util.go
+++ b/pkg/sql/exec_util.go
@@ -771,6 +771,10 @@ type ExecutorTestingKnobs struct {
 	// OnTempObjectsCleanupDone will trigger when the temporary objects cleanup
 	// job is done.
 	OnTempObjectsCleanupDone func()
+
+	// WithStatementTrace is called after the statement is executed in
+	// execStmtInOpenState.
+	WithStatementTrace func(span opentracing.Span, stmt string)
 }
 
 // PGWireTestingKnobs contains knobs for the pgwire module.

--- a/pkg/sql/testdata/ddl_analysis/ddl_analysis
+++ b/pkg/sql/testdata/ddl_analysis/ddl_analysis
@@ -1,0 +1,44 @@
+# The goal of these tests is to catch when a code change significantly
+# increases the number of round trips necessary to execute a given DDL
+# statement.
+#
+# NOTE: If the count (number of kv batch operations) increase is expected or
+# insignificant or the count decreases, update the test counts.
+# The rewrite option can be used to overwrite all the counts.
+
+exec
+CREATE USER TEST
+----
+
+exec
+CREATE TABLE t0()
+----
+
+count
+GRANT ALL ON * TO TEST
+----
+10
+
+
+exec
+CREATE TABLE t1();
+CREATE TABLE t2();
+CREATE TABLE t3();
+CREATE TABLE t4();
+CREATE TABLE t5();
+CREATE TABLE t6();
+CREATE TABLE t7();
+CREATE TABLE t8();
+CREATE TABLE t9();
+CREATE TABLE t10();
+----
+
+count
+GRANT ALL ON * TO TEST
+----
+160
+
+count
+CREATE ROLE rolea
+----
+6

